text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,402 @@
+
+`cargo` does not provide them for regular (untemplated) `CARGO_TARGET_DIR`. This is not a limitation when using the environment variable set globally, since all processes can read it, but it is one when this config is only set on specific calls or via `target-dir` in the config, meaning others tools cannot easily pick it up (and most external tools don't use `cargo-metadata`, which makes them all broken by default, but fixing this situation is not this RFC's purpose).
+
+When a templated `CARGO_TARGET_DIR` is used (in any form) (and not overriden, for example a templated env var overriden by a precise `--target-dir` option), it *will* use a forward link by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if the templated `CARGO_TARGET_DIR` wasn't set at all.
+
+This has a two big advantages: not breaking external tools and giving easy access to artifacts produced by `cargo build/test/doc` to users (they're in the habit of typing `./target/debug/my-bin`, this would continue working with forward links).
+
+A config option (CLI, `config.toml` and env var), `link-target-dir`, will be introduced to deactivate this behaviour, but it will `true` by default, for the reasons provided in favor of forward links just above.
+
+#### Detailed working of forward links
+
+When creating a forward link `cargo` will first attempt to create a symbolic link (regardless of the platform). If that fails, it will attempt zero or more platform-specific solutions, like junction points on NTFS. If that fails too, a warning will be emitted but this will not prevent the rest of the action to go on: regular calls like `cargo check/clippy/build/test` likely won't need this forward link and after the user has been warned they could either resolve the problem themselves or ignore it, depending on their own use case and domain-specific knowledge.
+
+### Providing backlinks
+
+Backlinks are metadata in templated `CARGO_TARGET_DIR` that links target directories back to the workspace they came from.
+
+[`targo`][tg] uses them in its own form of the feature and `cargo` uses them too for templated `CARGO_TARGET_DIR`.
+
+While details of the stored data are left to the implementation (there is no need for `cargo` to expose this data directly, though it could be exposed through `cargo metadata` in the future, see the relevant section below), one could imagine using it to clean target directories whose corresponding workspace does not exist anymore when calling something like `cargo clean --all-workspaces` (doing it automatically is not possible, else any workspace on external disks would have its target directory cleaned up each time the disk is unmounted, which is way too aggressive a default).
+
+# Drawbacks
```

Comment:
For `--target-dir`, some shells use `{}` and this would need escaping or quoting to be used on the command line

Reply:
I can't find a single mainstream shell where that's the case baring those that also interpret python ?

- Bash, Fish & ZSH use `${}`, not just `{}`
- Nushell uses `$"()"`

Reply:
Bash has [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html).

However, it looks like if a comma or a sequence is not present, then it will be treated as a literal. For with a cargo-cult familiarity of brace expansion, like me, we should call it out and that so long as a variable doesn't have commas or `..`, we should not run into problems with brace expansion.

Reply:
Oh I completely forgot about that, you're right.

Quick tests on macOS with Bash and ZSH show that simply wrapping in `""` also works so it should be fairly easy to work around.

```bash
cd /tmp
mkdir tests

ls ./{tests,}   # SUCCESS

ls ./{tests}    # FAIL
ls "./{tests}"  # FAIL
ls "./{tests,}" # FAIL
```

Done in https://github.com/rust-lang/rfcs/pull/3371/commits/025b8effa206b047a18fceca2661db1ec4c4d554